File: src/netpbm/pgm.c
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <ctype.h>
#include <arpa/inet.h>

#include "netpbm/netpbm.h"
#include "netpbm/pgm.h"

void print_netpgm(NETPBM *p)
{
    uint8_t *data = (uint8_t *) p->data;
    for (size_t i = 0; i < p->height; i++)
    {
        for (size_t j = 0; j < p->width; j++)
        {
            size_t pix = 0;

            if (p->bit_depth == 16)
            {
                // clamping to 0-255, but still supporting 0-65535
                pix = ntohs(((uint16_t *) data)[i * p->width + j]) * 65535 / p->max_value / 255;
            }
            else
                pix = data[i * p->width + j] * 255 / p->max_value;

            printf("\033[48;2;%lu;%lu;%lum  \033[m", pix, pix, pix);
        }
        puts("");
    }
}

File: src/parser/netpbm/pgm.rs
use std::io::Write;

use anyhow::Result;

use crate::{color::Color, pxbm_write, pxbm_writeln};

use super::Netpbm;

impl Netpbm {
    pub fn print_netpgm(&self, writer: &mut dyn Write) -> Result<()> {
        for i in 0..self.header.height {
            for j in 0..self.header.width {
                let pix = if self.header.bit_depth == 16 {
                    self.data[(i * self.header.width + j) as usize] * 65535 / self.header.max_value / 255
                } else {
                    self.data[(i * self.header.width + j) as usize] * 255 / self.header.max_value
                };

                pxbm_write!(writer, "{}", Color::new((pix as u8, pix as u8, pix as u8)))?;
            }
            pxbm_writeln!(writer)?;
        }
        Ok(())
    }
}